xocto/fields/ranges.py

```diff
@@ -0,0 +1,68 @@
+"""
+Adapter fields for postgres range types that return xocto.Range python objects.
+"""
+import datetime
+import json
+from typing import Optional
+
+from django.contrib.postgres import fields as pg_fields
+from django.contrib.postgres.fields import ranges as pg_ranges
+from django.contrib.postgres.fields import utils as pg_utils
+from django.db import models
+
+from xocto import ranges
+
+
+class FiniteDateRangeField(pg_fields.DateRangeField):
+    """
+    A DateRangeField with Inclusive-Inclusive [] bounds.
+
+    The underlying postgres type will always store the range as [). This field
+    will always translate to [].
+
+    Accepts and returns xocto.ranges.FiniteDateRange objects.
+    """
+
+    def get_prep_value(
+        self, value: Optional[ranges.FiniteDateRange]
+    ) -> Optional[ranges.FiniteDateRange]:
+        if value is None:
+            return None
+        return pg_ranges.DateRange(lower=value.start, upper=value.end, bounds="[]")
+
+    def from_db_value(
+        self,
+        value: Optional[pg_ranges.DateRange],
+        expression: object,
+        connection: object,
+    ) -> Optional[ranges.FiniteDateRange]:
+        if value is None:
+            return None
+        # DateRange is always stored as [) in the database, so we need to remove
+        # a day from the upper bound to make it inclusive.
```

Comment:
I don't fully understand whether that would have undesirable consequences or not, is there more that can be added to this comment to explain the choice of a day?

Reply:
I can expand on the comment, yes. But, for your understanding prior to that (and using numbers for terseness):

If we have `[1, 10]`, postgres will store that as `[1, 11)` AND return it as `[1, 11)`. We deduct 1 to get back to [1, 10]`. This is how all discrete ranges work in postgres https://www.postgresql.org/docs/current/rangetypes.html#RANGETYPES-DISCRETE

Reply:
and the range is always in full days?

Reply:
DateRange is, yes 👍 

DateTimeRanges are continuous and used with `[)` bounds, so there's no need for canonicalisation